glean_parser/javascript_server.py

```diff
@@ -0,0 +1,164 @@
+
+Warning: this outputter supports limited set of metrics.
+"""
+from collections import defaultdict
+from pathlib import Path
+from typing import Any, Dict, Optional, List
+
+from . import __version__
+from . import metrics
+from . import util
+
+
+def event_class_name(pingName: str) -> str:
+    return util.Camelize(pingName) + "ServerEvent"
+
+
+def generate_metric_name(metric: metrics.Metric) -> str:
+    return f"{metric.category}.{metric.name}"
+
+
+def generate_metric_argument_name(metric: metrics.Metric) -> str:
+    return f"{metric.category}_{metric.name}"
+
+
+def generate_js_metric_type(metric: metrics.Metric) -> str:
+    # TODO: this works with strings, implement other types
+    return metric.type
+
+
+def generate_metric_argument_description(metric: metrics.Metric) -> str:
+    return metric.description.replace("\n", " ").rstrip()
+
+
+def generate_ping_factory_method(ping: str) -> str:
+    return f"create{util.Camelize(ping)}EventFn"
+
+
+def output(
+    lang: str,
+    objs: metrics.ObjectTree,
+    output_dir: Path,
+    options: Optional[Dict[str, Any]] = None,
+) -> None:
+    """
+    Given a tree of objects, output Javascript or Typescript code to `output_dir`.
+
+    :param lang: Either "javascript" or "typescript";
+    :param objects: A tree of objects (metrics and pings) as returned from
+        `parser.parse_objects`.
+    :param output_dir: Path to an output directory to write to.
+    """
+
+    template = util.get_jinja2_template(
+        "javascript_server.jinja2",
+        filters=(
+            ("event_class_name", event_class_name),
+            ("metric_name", generate_metric_name),
+            ("metric_argument_name", generate_metric_argument_name),
+            ("js_metric_type", generate_js_metric_type),
+            ("metric_argument_description", generate_metric_argument_description),
+            ("factory_method", generate_ping_factory_method),
+        ),
+    )
+
+    # In this environment we don't use a concept of measurement window for collecting
+    # metrics. Only "events as pings" are supported.
+    # For each ping we generate code which contains all the logic for assembling it
+    # with metrics, serializing, and submitting. Therefore we don't generate classes for
+    # each metric as in standard outputters.
+    if "pings" not in objs:
+        print(
+            "❌ No ping definition found. Server-side environment is simplified and this"
+            + "parser doesn't generate individual metric files."
+        )
+        return
+
+    # go through all metrics in objs and build a map of
+    # ping->list of metric categories->list of metrics
+    ping_to_metrics: Dict[str, Dict[str, List[metrics.Metric]]] = defaultdict(dict)
+    for _category_key, category_val in objs.items():
+        for _metric_name, metric in category_val.items():
+            if isinstance(metric, metrics.Metric):
+                if metric.type not in ["string"]:
```

Comment:
It might make sense to pull out this list into a constant for easier modification later. And this would need some docs on when and how we can extend that